titus_isolate/event/oversubscribe_event_handler.py

```diff
@@ -104,72 +105,74 @@ def _handle(self, event):
+            simple_cpu_prediction = simple_cpu_usage_predictions.get(workload.get_id(), None)
+            if simple_cpu_prediction is None:
+                log.warning("No CPU prediction for workload: %s", workload.get_id())
+                continue
 
-                if pred_usage > threshold:
-                    log.info("Workload: %s is NOT oversubscribable: %s", workload.get_id(), pred_usage)
-                    continue
+            # Process prediction
+            pred_usage = simple_cpu_prediction / workload.get_thread_count()
+            threshold = self.__config_manager.get_float(TOTAL_THRESHOLD, DEFAULT_TOTAL_THRESHOLD)
 
-                log.info("Workload: %s is oversubscribable: %s", workload.get_id(), pred_usage)
-
-                if workload.is_opportunistic():
-                    # only add the number of "real" threads (non-opportunistic)
-                    free = workload.get_thread_count() - workload.get_opportunistic_thread_count()
-                    if free <= 0:
-                        continue
-                    underutilized_cpu_count += free
-                else:
-                    underutilized_cpu_count += workload.get_thread_count()
-                workload_count += 1
-
-            free_cpu_count = underutilized_cpu_count
-            if free_cpu_count > 0:
-                self.__window_publisher.add_window(start, end, free_cpu_count)
-
-            self.__success_count += 1
-            self.handled_event(event,
-                               'oversubscribed {} cpus from {} workloads, {} total cpus are oversubscribed'.format(
-                                   free_cpu_count, workload_count, underutilized_cpu_count))
-        except:
-            self.__fail_count += 1
-            log.exception("Event handler: '{}' failed to handle event: '{}'".format(
-                self.__class__.__name__, event))
+            log.info("Testing oversubscribability of workload: {}, threshold: {}, prediction: {}".format(
+                workload.get_id(), threshold, pred_usage))
+
+            if pred_usage > threshold:
+                log.info("Workload: %s is NOT oversubscribable: %s", workload.get_id(), pred_usage)
+                continue
+
+            log.info("Workload: %s is oversubscribable: %s", workload.get_id(), pred_usage)
+
+            if workload.is_opportunistic():
+                # only add the number of "real" threads (non-opportunistic)
+                free = workload.get_thread_count() - workload.get_opportunistic_thread_count()
+                if free <= 0:
+                    continue
+                underutilized_cpu_count += free
+            else:
+                underutilized_cpu_count += workload.get_thread_count()
+            workload_count += 1
+
+        free_cpu_count = underutilized_cpu_count
+        if free_cpu_count > 0:
+            self.__window_publisher.add_window(start, end, free_cpu_count)
+            self.__window_end_time = end
```

Comment:
Should we be holding the window lock here?

Reply:
Window publishing is serialized through the oversubscribe_event_handler. There are no concurrency protections in the publisher any more.

Reply:
Nevermind.  It's already held.